virttest/libvirt_vm.py

```diff
@@ -1467,10 +1469,11 @@ def _create_serial_console(self):
                                                        prompt=prompt)
             if not self.serial_console.is_alive():
                 LOG.error("Failed to create serial_console")
-            # Cause serial_console.close() to close open log file
-            self.serial_console.set_log_file(output_filename)
-            self.serial_console_log = os.path.join(utils_misc.get_log_file_dir(),
+            self.serial_console_log = os.path.join(utils_logfile.get_log_file_dir(),
                                                    output_filename)
+            # Cause serial_console.close() to close open log file
+            self.serial_console.set_log_file(self.serial_console_log)
+            self.serial_console.close_hooks += [utils_logfile.close_own_log_file]
```

Comment:
Okay, so to my understanding the key change of this proposal is that (logically speaking) it replaced the close hook with a VT defined one, in order to get rid of tracking the log handlers via a global variant managed by both aexpect and VT.

However I have the concern that I'm not sure if the `Spawn.close_hooks` was designed as a public attribute for endusers, since there was the `_add_close_hook` method as the setter which decorated with the leading `_`. IMO, using a private attribute/method for the replacement is just yet another workaround.

I think the critical point of this issue is about to determine either of aexpect or VT should own the log handlers by itself but not both. And in the case of we'd prefer to let VT take the ownership, the lifecycle of the log handlers would be decoupled from the session instances therefore a close hook would not always be a must to be set. On the other hand, aexpect could take over the thing as well, with the registration of the log handlers to be integrated into the `set_log_file` method.

Reply:
Let me add some quick comments here:

> However I have the concern that I'm not sure if the Spawn.close_hooks was designed as a public attribute for endusers, since there was the _add_close_hook method as the setter which decorated with the leading _. IMO, using a private attribute/method for the replacement is just yet another workaround.

Since we have a public `Spawn.close_hooks` list and not a private one like `Spawn._close_hooks` we could also assume that the hooks have always been meant to be public and customizable. The way a close hook was added privately before should not impeded that.

> I think the critical point of this issue is about to determine either of aexpect or VT should own the log handlers by itself but not both. And in the case of we'd prefer to let VT take the ownership, the lifecycle of the log handlers would be decoupled from the session instances therefore a close hook would not always be a must to be set. On the other hand, aexpect could take over the thing as well, with the registration of the log handlers to be integrated into the set_log_file method.

Indeed, there are some quite important comments on the [aexpect side](https://github.com/avocado-framework/aexpect/pull/124#discussion_r1379734217) of this where I would like to hear your opinion as well in order to conclude something more definite here. Do you agree with the idea to concentrate all logging functionality of aexpect via an output function there?

Reply:
> Since we have a public Spawn.close_hooks list and not a private one like Spawn._close_hooks we could also assume that the hooks have always been meant to be public and customizable.

Well, as we can see there is almost not any private attribute been defined by the aexpect's classes while some of them wasn't that meaningful to be accessed publicly. Anyway, I am not that keen on the rectification of that thing so it's fine for me to accept the current approach even though it was not the one I'd advocate.

> Do you agree with the idea to concentrate all logging functionality of aexpect via an output function there?

Yeah, I'd think so since the `output_func` hook was designed in a complete and self-consistent behavior. It already gives the users the maximized flexibility to handle the line-buffered output of an execution (for whatever case of logging, monitoring, etc.), therefore I didn't see the need to replace it with something else as VT has already chosen the use.

Reply:
> > Since we have a public Spawn.close_hooks list and not a private one like Spawn._close_hooks we could also assume that the hooks have always been meant to be public and customizable.
> 
> Well, as we can see there is almost not any private attribute been defined by the aexpect's classes while some of them wasn't that meaningful to be accessed publicly. Anyway, I am not that keen on the rectification of that thing so it's fine for me to accept the current approach even though it was not the one I'd advocate.

@ldoktor What do you think about this? At least having a public hooks list that one can add hooks to seems reasonable enough to me. @luckyh Any specific suggestions on alternative solution?

> > Do you agree with the idea to concentrate all logging functionality of aexpect via an output function there?
> 
> Yeah, I'd think so since the `output_func` hook was designed in a complete and self-consistent behavior. It already gives the users the maximized flexibility to handle the line-buffered output of an execution (for whatever case of logging, monitoring, etc.), therefore I didn't see the need to replace it with something else as VT has already chosen that use.

Ok, so I think we are all for having just the most flexible and powerful `output_func` approach to logging and any kind of output handling by aexpect. I guess if we agree on the current approach here we could look into the aexpect side about this next. So far the sibling PR ended with me mentioning about my worry about breaking changes [here](https://github.com/avocado-framework/aexpect/pull/124#discussion_r1379734217).

Reply:
I don't see anything wrong in adding arbitrary close hooks to `aexpect`, we do our best to execute them so as long as they are reasonable (which is on the users) it should be safe.

Reply:
> Any specific suggestions on alternative solution?

Ok, since now the log handlers are managed by VT completely, it would be possible to move the cleanup procedure into the same place where the sessions being closed.

https://github.com/avocado-framework/avocado-vt/blob/60ef8ea2e02bf7f824ee4428689828d1e51f6d75/virttest/env_process.py#L626

On the other side, we could consider the way about doing the cleanup via aexpect's close hooks, however, what I'd suggest there would be to have a fix or at least an enhancement for introducing something like `add_close_hook`/`on_close` in stead of manipulating the `close_hooks` list directly, which can prevent the users doing harmful things.

Reply:
Hi @luckyh I am not sure I fully understand the alternative suggestion here.

> Ok, since now the log handlers are managed by VT completely, it would be possible to move the cleanup procedure into the same place where the sessions being closed.

Do you have a precise diff in mind so that I can better understand what you mean?

> what I'd suggest there would be to have a fix or at least an enhancement for introducing something like add_close_hook/on_close in stead of manipulating the close_hooks list directly, which can prevent the users doing harmful things.

As this is python I doubt we can prevent the users from doing harmful things no matter what we try. Also I find the replacement of a public `session.close_hooks.append(new_hook)` with `session.add_close_hook(new_hook)` to be more like replacing one already available public function with another one unless we really have some useful logic to add to the second case.

Reply:
Sorry for the late reply.

> Do you have a precise diff in mind so that I can better understand what you mean?

It would be something like the following.

```patch
diff --git a/virttest/env_process.py b/virttest/env_process.py
index 156bb376..7eb43c00 100644
--- a/virttest/env_process.py
+++ b/virttest/env_process.py
@@ -44,6 +44,7 @@ from virttest import utils_qemu
 from virttest import migration
 from virttest import utils_kernel_module
 from virttest import arch
+from virttest import utils_logfile
 from virttest.utils_conn import SSHConnection
 from virttest.utils_version import VersionInterval
 from virttest.staging import service
@@ -630,6 +631,8 @@ def postprocess_vm(test, params, env, name):
         except Exception:
             pass
 
+    utils_logfile.close_all_log_files()
+
     if params.get("vm_extra_dump_paths") is not None:
         vm_extra_dumps = os.path.join(test.outputdir, "vm_extra_dumps")
         if not os.path.exists(vm_extra_dumps):
diff --git a/virttest/utils_logfile.py b/virttest/utils_logfile.py
index bd915e8a..6ae10060 100644
--- a/virttest/utils_logfile.py
+++ b/virttest/utils_logfile.py
@@ -160,3 +160,16 @@ def close_log_file(filename):
                 _open_log_files.pop(key_to_remove)
     finally:
         _log_lock.release()
+
+
+def close_all_log_files():
+    global _open_log_files, _log_file_dir, _log_lock
+    if not _acquire_lock(_log_lock):
+        raise LogLockError("Could not acquire exclusive lock to access"
+                           " _open_log_files")
+    try:
+        while _open_log_files:
+            _, log_fd = _open_log_files.popitem():
+            log_fd.close()
+    finally:
+        _log_lock.release()
```

> As this is python I doubt we can prevent the users from doing harmful things no matter what we try.

The important thing is that it can give a warning to the reviewers (no matter human or bot) for doing a check on the "defective code" when they are reading about that.

Let's just recall on how this issue was raised. The story was starting from a suspicious data access was found, right?

> Also I find the replacement of a public `session.close_hooks.append(new_hook)` with `session.add_close_hook(new_hook)` to be more like replacing one already available public function with another one unless we really have some useful logic to add to the second case.

The key point is with accessing the `close_hooks` attribute directly one could do something like `session.close_hooks.clear()` which may not be allowed to be issued by a external caller, while there is no way for the latter to do as so.

Reply:
> It would be something like the following.

I am not sure if this is taking this PR too far already considering its objective of just separating the dependencies but I pushed your required change with some improved reusability of already existing code.

> The key point is with accessing the `close_hooks` attribute directly one could do something like `session.close_hooks.clear()` which may not be allowed to be issued by a external caller, while there is no way for the latter to do as so.

I don't see anything wrong with a user calling `session.close_hooks.clear()` as the close hooks attribute is meant to be public and provide access to all hooks and their overall management via the list. So I don't see why an external caller should not be allowed to issue this call or how they would be mislead in issuing such a call via the API.

Reply:
> Really? With manipulating the low-level abstraction, I mean the python built-in `list` stuff you mentioned above, users could even store and fetch pure text/binary data and then building up a brand new message bus subsystem,

This sounds more like having issues with the way python is made (which is excessively free and lacking validation I agree) and arguing for putting extra effort to fix it ourselves while in fact a lot of the language implicitly heavily relies on user, bot, and overall caller sanity.

Of course, adding extra validation and restricting usage ourselves is always welcome as long as this is easily feasible. But I am afraid right now I am out of resources for establishing a complete wrapper around the hooks list having `add_close_hook`, `remove_close_hook`, various validation of what callables are added, removed, inserted, sorted, and so on. This pull request has already taken months of reviews and further modifications I constantly have to come back to (switch context, catch up with what was done and what is the new thing that has to be done) and all I wanted was to quickly help out separating the two projects. So unless you have a specific minimal and very compact diff I could quickly adapt here that achieves and communicates what you want unfortunately I won't be able to help out much further than this.